popde/adaptive_kde.py

```diff
@@ -224,6 +224,143 @@ def optimize_parameters(self, method='loo_cv', fom_plot_name=None):
+            alpha (float, optional): Initial alpha value, default is 0.0.
+            dim_names (list, optional): Dimension names.
+            do_fit (bool, optional): Whether to fit the KDE upon initialization.
+            n_splits (int, optional): Number of splits for k-fold 
+               cross-validation, default is 2.
+        """
+
+        self.n_splits = n_splits
+        #not sure I need this
+        self.alpha = alpha
+        self.bandwidth = bandwidth
+
+        super().__init__(data, weights, input_transf, stdize, rescale, backend, 
+                bandwidth, alpha, dim_names, do_fit)
+
+    def loo_cv_score(self, rescale_factors_alpha):
+        """
+        Compute the Leave-One-Out Cross-Validation (LOO-CV) score
+          based on log-likelihood.
+
+        Args:
+            rescale_factors_alpha (array-like): Array of rescale factors per
+              dimension and alpha value.
+
+        Returns:
+            float: Negative sum of log-likelihood scores for LOO-CV.
+        """
+        alpha = rescale_factors_alpha[-1]
+        rescale_factor = rescale_factors_alpha[0:len(rescale_factors_alpha)-1]
+
+        from sklearn.model_selection import LeaveOneOut
+        loo = LeaveOneOut() 
+        fom = 0.
+        for train_index, test_index in loo.split(self.kde_data):
+            train_data, test_data = self.kde_data[train_index], self.kde_data[test_index]
+            local_weights = None # FIX ME
+            awkde = AdaptiveBwKDE(train_data, local_weights, input_transf=self.input_transf,
+                                  stdize=self.stdize, rescale=rescale_factor,
+                                  bandwidth=self.bandwidth, alpha=alpha)
+            fom += np.log(awkde.evaluate(test_data))
+        return -fom
+
+    def kfold_cv_score(self, rescale_factors_alpha, seed=42):
+        """
+        Perform k-fold cross-validation
+        """
+        alpha= rescale_factors_alpha[-1]
+        rescale_factor = rescale_factors_alpha[0:len(rescale_factors_alpha)-1]
+        from sklearn.model_selection import KFold
+        kf = KFold(n_splits=self.n_splits, shuffle=True, random_state=seed)
+        fom = []
+        for train_index, test_index in kf.split(self.kde_data):
+            train_data, test_data = self.kde_data[train_index], self.kde_data[test_index]
+            local_weights = None # FIX ME
+            awkde = AdaptiveBwKDE(train_data, local_weights, input_transf=self.input_transf,
+                                  stdize=self.stdize, rescale=rescale_factor,
+                                  bandwidth=self.bandwidth, alpha=alpha)
+            log_kde_eval = np.log(awkde.evaluate(test_data))
+            fom.append(log_kde_eval.sum())
+        return -sum(fom)
+
+    def optimize_rescale_parameters(self, initial_rescale_factor, initial_alpha=0.0, method='loo_cv', fom_plot_name=None, bounds=None):
+
+        """
+        Fixed bounds on rescale factor otehrwise it is not working
```

Comment:
I don't understand this comment. If you need to discuss whether the method works or gives good results please do so on slack or in the comments to this PR, not in code comments.

Reply:
done